### PR TITLE
x/gamm: Internal audit and documentation

### PR DESCRIPTION
**Pull request by AlpinYukseloglu**
Component of: #1835

## What is the purpose of the change

Ongoing audit of x/gamm module for all AMM-related logic.

Current progress is being tracked in [this document](https://docs.google.com/document/d/1Qz-2Gy2YMQqb_REJ1sC-u7jcG-9-exhM_yFD0S3Nh3Q/edit#).

## Brief Changelog

WIP, will add when complete.

## Testing and Verifying

This change added tests and can be verified as follows:
- WIP, will add when complete.

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable

### CODE REVIEW

File: x/gamm/keeper/genesis.go [hunk 1]
```diff
@@ -8,11 +8,13 @@ import (
...
 
 // InitGenesis initializes the capability module's state from a provided genesis
-// state.
+// state, which includes the current live pools, global pool parameters (e.g. pool creation fee), next pool number etc.
 func (k Keeper) InitGenesis(ctx sdk.Context, genState types.GenesisState, unpacker codectypes.AnyUnpacker) {
 	k.SetParams(ctx, genState.Params)
 	k.SetNextPoolNumber(ctx, genState.NextPoolNumber)
 
+	// Sums up the liquidity in all genesis state pools to find the total liquidity across all pools
+	// Also adds each genesis state pool to the capability module's state
```
Comment (mattverse) [inline x/gamm/keeper/genesis.go:17]: Srry if im missing context, what do you mean by capabilty module's state?

File: x/gamm/keeper/multihop.go [hunk 2]
```diff
@@ -68,7 +83,11 @@ func (k Keeper) MultihopSwapExactAmountOut(
 	return tokenInAmount, nil
 }
 
-// TODO: Document this function.
+// createMultihopExpectedSwapOuts defines the output denom and output amount for the last pool in
+// the route of pools the caller is intending to hop through in a fixed-output multihop tx. It estimates the input
+// amount for this last pool and then chains that input as the output of the previous pool in the route, repeating
+// until the first pool is reached. It returns an array of inputs, each of which correspond to a pool in the
```
Comment (mattverse) [inline x/gamm/keeper/multihop.go:89]:
```suggestion
// until the first pool is reached. It returns an array of inputs, each of which correspond to a pool ID in the
```

**Review by mattverse — APPROVED** [review]
Overall LGTM! Left some minor comments, please take a look

File: x/gamm/keeper/genesis.go [hunk 3]
```diff
@@ -8,11 +8,13 @@ import (
 )
 
 // InitGenesis initializes the capability module's state from a provided genesis
```
Comment (alexanderbez) [inline x/gamm/keeper/genesis.go:10]:
```suggestion
// InitGenesis initializes the x/gamm module's state from a provided genesis
```

File: x/gamm/keeper/genesis.go [hunk 4]
```diff
@@ -8,11 +8,13 @@ import (
...
-// state.
+// state, which includes the current live pools, global pool parameters (e.g. pool creation fee), next pool number etc.
 func (k Keeper) InitGenesis(ctx sdk.Context, genState types.GenesisState, unpacker codectypes.AnyUnpacker) {
 	k.SetParams(ctx, genState.Params)
 	k.SetNextPoolNumber(ctx, genState.NextPoolNumber)
 
+	// Sums up the liquidity in all genesis state pools to find the total liquidity across all pools
```
Comment (alexanderbez) [inline x/gamm/keeper/genesis.go:16]:
```suggestion
	// Sums up the liquidity in all genesis state pools to find the total liquidity across all pools.
```

File: x/gamm/keeper/multihop.go [hunk 5]
```diff
@@ -42,24 +47,34 @@ func (k Keeper) MultihopSwapExactAmountOut(
...
 
 	insExpected[0] = tokenInMaxAmount
 
+	// Iterates through each routed pool and executes their respective swaps. Note that all of the work to get the return
+	// value of this method is done when we calculate insExpected – this for loop primarily serves to execute the actual
+	// swaps on each pool.
 	for i, route := range routes {
 		_tokenOut := tokenOut
+
+		// If there is one pool left in the route, set the expected output of the current swap to the estimated input of the final pool
```
Comment (alexanderbez) [inline x/gamm/keeper/multihop.go:64]:
```suggestion
		// If there is one pool left in the route, set the expected output of the current swap
		// to the estimated input of the final pool.
```

File: x/tokenfactory/keeper/admins_test.go [hunk 6]
```diff
@@ -71,9 +71,7 @@ func (suite *KeeperTestSuite) TestAdminMsgs() {
 // * Only the admin of a denom can mint tokens for it
 // * The admin of a denom can mint tokens for it
 func (suite *KeeperTestSuite) TestMintDenom() {
-	var (
-		addr0bal = int64(0)
-	)
+	addr0bal := int64(0)
```
Comment (alexanderbez) [inline x/tokenfactory/keeper/admins_test.go:74]:
```suggestion
	var addr0bal int64
```

File: x/tokenfactory/keeper/admins_test.go [hunk 7]
```diff
@@ -122,9 +120,7 @@ func (suite *KeeperTestSuite) TestMintDenom() {
 }
 
 func (suite *KeeperTestSuite) TestBurnDenom() {
-	var (
-		addr0bal = int64(0)
-	)
+	addr0bal := int64(0)
```
Comment (alexanderbez) [inline x/tokenfactory/keeper/admins_test.go:123]:
```suggestion
	var addr0bal int64
```

File: x/gamm/keeper/pool.go [hunk 8]
```diff
@@ -43,7 +43,7 @@ func (k Keeper) GetPoolAndPoke(ctx sdk.Context, poolId uint64) (types.PoolI, err
 	return pool, nil
 }
 
-// Get pool, and check if the pool is active / allowed to be swapped against
+// Get pool, and check if the pool is active (allowed to be swapped against)
```
Comment (alexanderbez) [inline x/gamm/keeper/pool.go:46]:
```suggestion
// Get pool and check if the pool is active, i.e. allowed to be swapped against.
```

**Review by alexanderbez — COMMENTED** [review]
Nice work

**Review by ValarDragon — APPROVED** [review]
This LGTM, lets have the subsequent work & followups we discussed on Monday in additional PRs?

File: x/gamm/keeper/genesis.go (hunk 1 above)
Comment (AlpinYukseloglu) [reply x/gamm/keeper/genesis.go]: In this case it refers to the gamm module. Updated in recent commit to be clearer

**Pull request merged**